### PR TITLE
gh-113317: Argument Clinic: don't use global state in warn() and fail()

### PR DESCRIPTION
**Pull request by erlend-aasland**
Competing PR to #115371, #115370, and #115369

* Issue: gh-113317

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -6174,7 +6176,11 @@ def do_post_block_processing_cleanup(self, lineno: int) -> None:
             return
 
         self.check_remaining_star(lineno)
-        self.function.docstring = self.format_docstring()
+        try:
+            self.function.docstring = self.format_docstring()
+        except ClinicError as exc:
+            exc.lineno = lineno
```
Comment (sobolevn) [inline Tools/clinic/clinic.py:6182]: Question: why `filename` is not set here?

Comment (erlend-aasland) [reply Tools/clinic/clinic.py:6182]: I don't think it is needed; but OTOH, it would not hurt.

Comment (erlend-aasland) [reply Tools/clinic/clinic.py:6182]: Added in 3775ab38f90ae447466ee6cdead8777f4c3be29d

**Review by sobolevn — APPROVED** [review]
Thank you!

**Comment by erlend-aasland** [comment]
Thanks for the review!

**Pull request merged**